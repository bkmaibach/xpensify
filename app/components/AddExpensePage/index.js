import React from 'react';
import ExpenseForm from "../ExpenseForm"
import { connect } from 'react-redux';
import { addExpense } from "../../actions/expenses";

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    // props.dispatch(addExpense(expense));
    this.props.addExpense(expense);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
});

// The second function param is called mapDispatchToProps and is needed because
// of the difficulty in sending a spy in place of addExpense
// (addExpense comes from an imoprt, not a prop)
// It basically turns two(functions(x)) into oneFunction(x)
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
