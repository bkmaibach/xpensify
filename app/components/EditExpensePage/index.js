import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from '../ExpenseForm';
import { editExpense, removeExpense } from "../../actions/expenses";

// Refactor as a class based component
// Map dispatch to props
export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
      this.props.editExpense(this.props.expense.id, expense);
      this.props.history.push('/');
      // console.log('updated', expense);
  }

  onClickRemove = (e) => {
    this.props.removeExpense({id: this.props.expense.id});
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button
          onClick={this.onClickRemove}>Remove</button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find( (expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, updates) => dispatch(editExpense(id, updates)),
  removeExpense: ({id}) => dispatch(removeExpense({id}))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
