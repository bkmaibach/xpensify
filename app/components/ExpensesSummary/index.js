import React from 'react';
import { connect } from 'react-redux';
import expensesTotalSelector from '../../selectors/expenses-total';
import expensesSelector from '../../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = ({count, total}) => (
  <div>
    {count === 0 ? (
      <h3>Add expenses to begin</h3>
    ) : (
    <h3>Viewing {count} expense{ count > 1 ? ('s') : ('') } totalling {numeral(total/100).format('$0,0.00')}</h3>
    ) }
  </div>
);

const mapStateToProps = (state) => {
  const selectedExpenses = expensesSelector(state.expenses, state.filters);
  return {
    total: expensesTotalSelector(selectedExpenses),
    count: selectedExpenses.length
  }
}

export default connect(mapStateToProps)(ExpensesSummary);