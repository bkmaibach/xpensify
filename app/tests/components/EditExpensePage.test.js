import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpense, removeExpense, expense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  expense = expenses[2];
  history = { push: jest.fn() };
  wrapper = shallow(
  <EditExpensePage
    editExpense={editExpense}
    removeExpense={removeExpense}
    expense={expense}
    history={history}
  />);
});

test('should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit for editing expense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expense)
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expense.id, expense);
});

test('should handle onClickRemove for removing expense', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith({id: expense.id});
});

// Should render the EditExpensePage
// Should handle editExpense (using spies)
// Should handle removeExpense (using spies)