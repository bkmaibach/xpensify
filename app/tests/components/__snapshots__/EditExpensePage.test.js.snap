// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 3000000000,
        "createdAt": 7000000000,
        "description": "trial2",
        "id": "2",
        "note": "note2",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
